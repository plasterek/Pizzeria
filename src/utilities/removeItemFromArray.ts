export const removeItemFromArray = <T>(array: T[], item: T): void => {
  const indexOfItem = array.findIndex((items) => items === item);
  array.splice(indexOfItem, 1);
};
