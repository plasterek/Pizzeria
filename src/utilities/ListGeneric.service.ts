import {
  IGenericObject,
  TGenericProperty,
  IGenericListObject,
} from "./models/Generics.model";
import { removeItemFromArray } from "./removeItemFromArray";

export class List<T extends IGenericObject, K extends TGenericProperty> {
  private list: IGenericListObject<T, K>[] = [];

  public add(object: T, property: K): void {
    this.list.push({ object, property });
  }
  public delete(object: T): void {
    const itemExist: IGenericListObject<T, K> | undefined = this.getObjectById(
      object.id
    );
    if (!itemExist) {
      throw new Error();
    }
    removeItemFromArray(this.list, itemExist);
  }
  getList(): IGenericListObject<T, K>[] {
    return this.list;
  }
  public getObjectById(objectId: string): IGenericListObject<T, K> | undefined {
    const objectExist: IGenericListObject<T, K> | undefined = this.list.find(
      (element) => element.object.id === objectId
    );
    return objectExist;
  }
}
