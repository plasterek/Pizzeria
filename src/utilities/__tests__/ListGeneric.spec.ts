import { List } from "../ListGeneric.service";
import { IGenericObject, TGenericProperty, IGenericListObject } from "../models/Generics.model";
import * as removeItem from "../../utilities/removeItemFromArray";

describe("List generic class", () => {
  //given
  const object: IGenericObject = { id: "id" };
  const property: TGenericProperty = "string";
  let list = new List();
  beforeEach(() => {
    list = new List();
    jest.clearAllMocks();
  });
  describe("When trying to add object with property", () => {
    it("It should appear in item list", () => {
      //when
      list.add(object, property);
      //then
      expect(list).toMatchSnapshot();
    });
  });
  describe("When trying to delete object and object does not exist in list", () => {
    it("It should throw an exception", () => {
      //then
      expect(() => list.delete(object)).toThrow();
    });
  });
  describe("When trying to delete object and everything went well", () => {
    it("getObjectById method should be caled", () => {
      //given
      const get = jest.spyOn(list, "getObjectById");
      //when
      list.add(object, property);
      list.delete(object);
      //then
      expect(get).toBeCalled();
    });
    it("Object should dissapear from list", () => {
      //when
      list.add(object, property);
      list.delete(object);
      //then
      expect(list).toMatchSnapshot();
    });
    it("removeItemFromArray function should be called", () => {
      //given
      const remove = jest.spyOn(removeItem, "removeItemFromArray");
      //when
      list.add(object, property);
      list.delete(object);
      //then
      expect(remove).toBeCalled();
    });
  });
  describe("When trying to get list", () => {
    it("It should return an array", () => {
      //then
      expect(list.getList()).toMatchObject([]);
    });
  });
  describe("When trying to get object by Id and object does not exist in list", () => {
    it("It should return undefined", () => {
      //then
      expect(list.getObjectById(object.id)).toBe(undefined);
    });
  });
  describe("When trying to get object by Id and everything goes well", () => {
    it("It should return given object with property", () => {
      const expectedResult: IGenericListObject<IGenericObject, TGenericProperty> = { object, property };
      //when
      list.add(object, property);
      //then
      expect(list.getObjectById(object.id)).toMatchObject(expectedResult);
    });
  });
});
