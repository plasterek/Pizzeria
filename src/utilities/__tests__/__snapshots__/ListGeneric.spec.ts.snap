// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List generic class When trying to add object with property It should appear in item list 1`] = `
List {
  "list": [
    {
      "object": {
        "id": "id",
      },
      "property": "string",
    },
  ],
}
`;

exports[`List generic class When trying to delete object and everything went well Object should dissapear from list 1`] = `
List {
  "list": [],
}
`;
